execution-engine/src/hcall/common.rs

```diff
@@ -1003,196 +361,124 @@ pub(crate) enum EntrySignature {
-
-    /// Returns the current lifecycle state that the host provisioning state is
-    /// in.
-    fn get_lifecycle_state(&self) -> LifecycleState;
-
-    /// Returns the current number of data sources provisioned into the host
-    /// provisioning state.
-    fn get_current_data_source_count(&self) -> usize;
-
-    /// Returns the expected data sources, as identified by their client IDs,
-    /// that we expect to be provisioned into the host state.
-    fn get_expected_data_sources(&self) -> Vec<u64>;
-
-    /// Returns the list of client IDs of clients who can request shutdown of
-    /// the platform.
-    fn get_expected_shutdown_sources(&self) -> Vec<u64>;
-
-    /// Returns the current number of stream sources provisioned into the host
-    /// provisioning state.
-    fn get_current_stream_source_count(&self) -> usize;
-
-    /// Returns the expected stream sources, as identified by their client IDs,
-    /// that we expect to be provisioned into the host state.
-    fn get_expected_stream_sources(&self) -> Vec<u64>;
-
-    /// Returns a result of a WASM computation that has executed on the host
-    /// provisioning state.  Returns `None` iff no such result has been
-    /// registered.
-    fn get_result(&self) -> Option<Vec<u8>>;
-
-    /// Returns an SHA-256 digest of the bytes loaded into the host provisioning
-    /// state.  Returns `None` iff no such program has yet been loaded.
-    fn get_program_digest(&self) -> Option<Vec<u8>>;
-
-    /// Sets the expected data sources, through a list of their source IDs, that
-    /// this computation is expecting.
-    fn set_expected_data_sources(&mut self, sources: &[u64]) -> &mut dyn ExecutionEngine;
-
-    /// Sets the expected stream sources, through a list of their source IDs, that
-    /// this computation is expecting.
-    fn set_expected_stream_sources(&mut self, sources: &[u64]) -> &mut dyn ExecutionEngine;
-
-    /// Sets the expected shutdown sources, through a list of their source IDs, that
-    /// this computation is expecting.
-    fn set_expected_shutdown_sources(&mut self, sources: &[u64]) -> &mut dyn ExecutionEngine;
-
-    /// Registers the previous result.
-    fn set_previous_result(&mut self, result: &Option<Vec<u8>>);
-
-    /// Moves the host provisioning state's lifecycle state into
-    /// `LifecycleState::Error`, a state which it cannot ever escape,
-    /// effectively invalidating it.
-    fn invalidate(&mut self);
-
-    /// Signals that a client would like to shutdown the platform.  Has no
-    /// effect is `client_id` does not correspond to a client with the shutdown
-    /// role.
-    fn request_shutdown(&mut self, client_id: u64);
-
-    ////////////////////////////////////////////////////////////////////////////
-    // Derived code.
-    ////////////////////////////////////////////////////////////////////////////
+    fn invoke_entry_point(&mut self, file_name: &str) -> Result<EngineReturnCode, FatalEngineError>;
```

Comment:
This can be made more general by writing:

```rust
fn invoke_entry_point<T>(&mut self, file_name: T) -> Result<EngineReturnCode, FatalEngineError>
where
  T: AsRef<str>;
```

Then `String`, `&String`, `&str`, and so on, can all be passed to the function without explicit conversion.

Reply:
Ah, I see.  That makes sense, I've hit this problem before, too.